noria-server/src/controller/inner.rs

```diff
@@ -1109,7 +1097,7 @@ impl ControllerInner {
         graphviz(&self.ingredients, detailed, &self.materializations)
     }
 
-    fn remove_leaf(&mut self, mut leaf: NodeIndex) -> Result<(), String> {
+    fn remove_query_node(&mut self, leaf: NodeIndex) -> Result<(), String> {
```

Comment:
@fintelia can you take a look at the change to this method? I'm not familiar with what it does and what invariants caller may assume.

Reply:
Ping @fintelia

Reply:
I don't think these changes are right. This code will panic if the node to be removed does not have exactly one child, but I think it should actually be asserting that the indicated node has no non-reader children. I'm also not sure I understand why the function name was changed.

Reply:
I changed the structure of the graph such that views are connected to readers through egress/ingress nodes, so these nodes shouldn't be directly connected to readers anymore.

I might have misunderstood the function purpose -- I changed the name since I thought removing a "leaf" implied the leaf had no children, which is not the case. Rather the node passed in is the query node with an egress node as a child (used to be a reader as a child). I'm not quite sure I understand the old comment about how the query leaf node may have "egress nodes or other, dependent queries" as children, but the code panicked if `has_non_reader_children` was true.

Reply:
I see, so the function _should_ only be called on leaf nodes of the MIR graph, and should panic otherwise. Sometimes those leaf nodes are query nodes and sometimes they are just other nodes being removed in topological order (no children). So I should modify the code to account for the second case. Does that sound correct?

Reply:
Yeah, that sounds right. You might also want to check with @ms705 to make sure I'm recalling the MIR graph part correctly.